src/backend/api/meals.js

```diff
@@ -2,14 +2,93 @@ const express = require("express");
 const router = express.Router();
 const knex = require("../database");
 
-router.get("/", async (request, response) => {
-  try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
-  } catch (error) {
-    throw error;
-  }
+//Returns all meal titles
+router.get("/titles", async(request, response) => {
+    try {
+        // knex syntax for selecting things. Look up the documentation for knex for further info
+        const titles = await knex("meals").select("title");
+        response.json(titles);
+    } catch (error) {
+        throw error;
+    }
+});
+//Adds a new meal
+router.post("/", async(request, response) => {
+    try {
+        const newMeal = await knex("meals").insert(request.body);
+        response.send(`New added meal: ${newMeal}`);
+    } catch (error) {
+        console.log(error);
+        response.send(error);
+    }
+});
+//Returns meal by id
+router.get("/:id", async(request, response) => {
+    try {
+        const mealById = await knex("meals").where("id", parseInt(request.params.id));
+        response.json(mealById);
+    } catch (error) {
+        console.log(error);
+        response.send(error);
+    }
+});
+//Updates the meal by id	
+router.put("/:id", async(request, response) => {
+    try {
+        const updateMeal = await knex("meals")
+            .where("id", parseInt(request.params.id))
+            .update(request.body);
+        response.json(updateMeal);
+    } catch (error) {
+        console.log(error);
+        response.send(error);
+    }
+});
+//Deletes the meal by id
+router.delete("/:id", async(request, response) => {
+    try {
+        const deletedMeal = await knex("meals")
+            .where("id", parseInt(request.params.id))
+            .del();
+        response.json(`Deleted meal: ${deletedMeal}`);
+    } catch (error) {
+        console.log(error);
+        response.send(error);
+    }
+});
+//Query Parameters
+router.get("/", async(request, response) => {
+    let meals = knex('meals');
+
+    if ('maxPrice' in request.query) {
+        meals = meals.where('price', '<=', parseInt(request.query.maxPrice))
+    }
+
+    if ('availableReservations' in request.query) {
+        const subquery = knex.select('*').from('reservations')
+        meals = meals.select("*")
+            .whereIn("id", subquery)
+            .having("meals.max_reservations", ">", "sum('reservations.number_of_guests')");
+    }
```

Comment:
So, you have the meals table, and each meal can have several reservations. This a one (meal) to-many (reservations) relationship.
One way to go with the availableReservations param is that you need to find out how many reservations a meal has, and then compare that with the max_reservations number.
I suggest you first create a query that does a left join with the reservations (this way you take into account meals that don't have even one reservation), count the reservations.meal_id and use having to return the meals ids with a higher max_reservations then the count. So, after you have the meals id you can just do a whereIn query.
I believe something close to this will do the job:
```
const availableMealsIds = knex('meal')
    .select('meal.id', 'meal.max_reservations')
    .leftJoin('reservation', { 'meal.id': 'reservation.meal_id' })
    .count({'reservations' : 'reservation.meal_id'})
    .groupBy('meal.id')
    .having(knex.raw('meal.max_reservations > reservations'));
```
And then, simply use these Ids to get the meals you need :)